apps/opentelemetry/src/otel_batch_olp.erl

```diff
@@ -0,0 +1,623 @@
+exporting(state_timeout, empty_table, Data) ->
+    {next_state, idle, Data};
+exporting(state_timeout, exporting_timeout, Data) ->
+    %% kill current exporting process because it is taking too long
+    Data1 = kill_runner(Data),
+    {next_state, idle, Data1};
+%% important to verify runner_pid and FromPid are the same in case it was sent
+%% after kill_runner was called but before it had done the unlink
+%% Exit reason is ignored, since we don't handle exporter failures specifically for now
+exporting(info, {'EXIT', FromPid, _}, Data=#data{runner_pid=FromPid}) ->
+    complete_exporting(Data);
+exporting(EventType, Event, Data) ->
+    handle_event_(exporting, EventType, Event, Data).
+
+terminate(_Reason, _State, #data{exporter=ignore}) ->
+    ok;
+terminate(_Reason, State, Data=#data{exporter=Exporter,
+                                     resource=Resource,
+                                     external_config=ExtConfig,
+                                     atomic_ref=AtomicRef,
+                                     tables={Tab1, Tab2},
+                                     shutdown_ms=ShutdownMs,
+                                     otel_signal=Signal
+                                    }) ->
+    ?disable(AtomicRef),
+    T0 = ?time_ms,
+    _ = maybe_wait_for_current_runner(State, Data, ShutdownMs),
+    T1 = ?time_ms,
+
+    %% Check both tables as each one may have some late unexported signals data.
+    %% NOTE: exports are attempted sequentially to follow the specification restriction:
+    %% "Export will never be called concurrently for the same exporter instance"
+    %% (see: https://opentelemetry.io/docs/specs/otel/logs/sdk/#export).
+    RemTime = ?rem_time(ShutdownMs, T0, T1),
+    ets:info(Tab1, size) > 0
+        andalso export_and_wait(Exporter, Resource, Tab1, ExtConfig, RemTime, Signal),
+    T2 = ?time_ms,
+    RemTime1 = ?rem_time(RemTime, T1, T2),
+    ets:info(Tab2, size) > 0
+        andalso export_and_wait(Exporter, Resource, Tab2, ExtConfig, RemTime1, Signal),
+
+    _ = otel_exporter:shutdown(Exporter),
+    ok.
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+handle_event_(_State, {call, From}, {change_config, NewConfig, NewExtConfig}, Data) ->
+    {keep_state,
+     add_mutable_config_to_data(NewConfig, NewExtConfig, Data),
+     [{reply, From, {ok, NewConfig}}]};
+handle_event_(_State, info, {'EXIT', Pid, Reason}, #data{runner_pid=RunnerPid})
+  when Pid =/= RunnerPid ->
+    %% This can be a linked exporter process, unless someone linked to the handler process,
+    %% or explicitly called exit(HandlerPid, Reason)
+    %% This will call terminate/3 and may try to export current log events,
+    %% even if the linked exporter process is down.
+    %% This is safe, though, as all errors of otel_exporter:export/5 are caught.
+    {stop, Reason};
+handle_event_(_State, _, _, _) ->
```

Comment:
Curious: do we purposely not postpone flush events when exporting?

Reply:
Yes, when the system is under constant stress, it would lead to accumulation of postponed events.